src/devices/Board/I2cDeviceManager.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Device.I2c;
+using System.Text;
+
+namespace Iot.Device.Board
+{
+    internal class I2cDeviceManager : I2cDevice
+    {
+        private readonly Board _board;
+        private readonly int _sdaPin;
+        private readonly int _sclPin;
+        private I2cDevice _i2cDeviceImplementation;
+
+        public I2cDeviceManager(Board board, I2cConnectionSettings settings, int[] pins, Func<I2cConnectionSettings, int[], I2cDevice> creationOperation)
+        {
+            _board = board;
+            _sdaPin = pins[0];
+            _sclPin = pins[1];
+            try
+            {
+                _board.ReservePin(_sdaPin, PinUsage.I2c, this);
+                _board.ReservePin(_sclPin, PinUsage.I2c, this);
+                _i2cDeviceImplementation = creationOperation(settings, pins);
+            }
+            catch (Exception)
+            {
+                _board.ReleasePin(_sdaPin, PinUsage.I2c, this);
+                _board.ReleasePin(_sclPin, PinUsage.I2c, this);
+                throw;
+            }
+        }
+
+        public override I2cConnectionSettings ConnectionSettings
+        {
+            get
+            {
+                return _i2cDeviceImplementation.ConnectionSettings;
+            }
+        }
+
+        internal I2cDevice RawDevice
+        {
+            get
+            {
+                return _i2cDeviceImplementation;
+            }
+        }
+
+        public override byte ReadByte()
+        {
+            return _i2cDeviceImplementation.ReadByte();
+        }
+
+        public override void Read(Span<byte> buffer)
+        {
+            _i2cDeviceImplementation.Read(buffer);
+        }
+
+        public override void WriteByte(byte value)
+        {
+            _i2cDeviceImplementation.WriteByte(value);
+        }
+
+        public override void Write(ReadOnlySpan<byte> buffer)
+        {
+            _i2cDeviceImplementation.Write(buffer);
+        }
+
+        public override void WriteRead(ReadOnlySpan<byte> writeBuffer, Span<byte> readBuffer)
+        {
+            _i2cDeviceImplementation.WriteRead(writeBuffer, readBuffer);
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
```

Comment:
if a device is recreated with same connection do we need to track some sort of reference counting?

Reply:
Normally, you would now create an I2cDevice for a particular address and then fed this into a binding (ie. a BME680, an MCP28xxx or whatever) and not directly use the device any further. If you create another instance of I2cDevice for the same physical device and use it differntly, or create another instance of a binding with it, you'll very likely run into weird problems, because the binding expects that any state change of the device only happens trough that binding. 

Reference counting would be needed if we want multiple instances for the same device, but I think we shouldn't allow that. So if we extend these checks, we should just prevent creation of the same I2cDevice a second time (until the first one is disposed). For that, remembering whether a particular device was already created is sufficient (can use a Dictionary or a Hashtable for that).